src/libraries/System.Threading.RateLimiting/src/System/Threading/RateLimiting/ChainedPartitionedRateLimiter.cs

```diff
@@ -0,0 +1,188 @@
+                }
+                catch (Exception ex)
+                {
+                    Exception? innerEx = CommonDispose(leases, i);
+                    if (innerEx is AggregateException aggregateException)
+                    {
+                        Exception[] exceptions = new Exception[aggregateException.InnerExceptions.Count + 1];
+                        aggregateException.InnerExceptions.CopyTo(exceptions, 0);
+                        exceptions[exceptions.Length - 1] = ex;
+                        throw new AggregateException(exceptions);
+                    }
+                    // REVIEW: Chose consistent exception type here, but could be convinced to just throw the original exception as is
+                    throw new AggregateException(ex);
+                }
+
+                if (!lease.IsAcquired)
+                {
+                    Exception? ex = CommonDispose(leases, i);
+                    return ex is not null ? throw ex : lease;
+                }
+
+                leases ??= new RateLimitLease[_limiters.Length];
+                leases[i] = lease;
+            }
+
+            return new CombinedRateLimitLease(leases!);
+        }
+
+        protected override async ValueTask<RateLimitLease> WaitAsyncCore(TResource resourceID, int permitCount, CancellationToken cancellationToken)
+        {
+            RateLimitLease[]? leases = null;
+            for (int i = 0; i < _limiters.Length; i++)
+            {
+                RateLimitLease lease;
+                try
+                {
+                    lease = await _limiters[i].WaitAsync(resourceID, permitCount, cancellationToken).ConfigureAwait(false);
+                }
+                catch (Exception ex)
+                {
+                    Exception? innerEx = CommonDispose(leases, i);
+                    if (innerEx is AggregateException aggregateException)
+                    {
+                        Exception[] exceptions = new Exception[aggregateException.InnerExceptions.Count + 1];
+                        aggregateException.InnerExceptions.CopyTo(exceptions, 0);
+                        exceptions[exceptions.Length - 1] = ex;
+                        throw new AggregateException(exceptions);
+                    }
+                    // REVIEW: Chose consistent exception type here, but could be convinced to just throw the original exception as is
+                    throw new AggregateException(ex);
+                }
+                if (!lease.IsAcquired)
+                {
+                    Exception? ex = CommonDispose(leases, i);
+                    return ex is not null ? throw ex : lease;
+                }
+
+                leases ??= new RateLimitLease[_limiters.Length];
+                leases[i] = lease;
+            }
+
+            return new CombinedRateLimitLease(leases!);
+        }
+
+        // REVIEW: Do we dispose the inner limiters? Or just mark the object as disposed and throw in all the methods
```

Comment:
I don't see why `ChainedPartitionedRateLimiter` should assume ownership of the inner limiters. It didn't create them. Throwing an ODE from all the methods after disposal seems fine.

The only thing I don't like about this is that the inner limiters might clean up outstanding leases when they're disposed, and this wouldn't do that. I don't want to be in the business of tracking outstanding leases and cleaning up ourselves at this layer though. We might ultimately have to do that here, but not right now.

For now, let's make sure we have clear documentation telling consumers to cancel all outstanding waits and dispose all outstanding leases before disposing the limiter returned from `CreateChained`. Or perhaps make that the guidance for `RateLimter`s and `PartitionedRateLimiter<TResource>`s in general.

That should be the case for our rate limiting middleware with graceful shutdown at least, but then there's always ungraceful shutdown where the host and the server timeout waiting for requests and dispose everything anyway 😞.